doc: GRASS Programming Style Guide (#3569)
OSGeo/grass: doc/development/style_guide.md

# Set the computational region with g.region as needed.
grass.run_command('g.region', raster='input')
gs.del_temp_region()
# Original region applies now.
```

This makes any changes done in the tool local for the tool without influencing
other tools running in the same session.

If you need even more control, use the GRASS_REGION environment variable which
is passed to subprocesses. Python API has functions which help with the setup:

```python
os.environ["GRASS_REGION"] = gs.region_env(raster=input_raster)
```

If different subprocesses need different regions, use different environments:

```python
env = os.environ.copy()
env["GRASS_REGION"] = gs.region_env(raster=input_raster)
gs.run_command("r.slope.aspect", elevation=input_raster, slope=slope, env=env)
```

This approach makes the computational region completely safe for parallel
processes as no region-related files are modified.

#### Temporary Maps

Using temporary maps is preferred over using temporary mapsets. This follows the
rule that writing should be done only to the current mapset. Some users may have
write permissions only for their mapsets, but not for creating other mapsets.

> The following script creates a temporary name using `gs.append_node_pid`` which

Reviewer: ```suggestion
The following script creates a temporary name using `gs.append_node_pid` which
```

Was the extra backtick intended (the PR syntax formatting was weird on my phone, maybe that was why)

Author: No, my editor is (partially) to blame

```

Tools need to accommodate input map names with (_elevation_) and without mapset
(_elevation@PERMANENT_). If you need only the map name without mapset, you can
do:

```python
map_name = map_name.split("@")[0]
```

If you need the full name or the mapset only, use _gs.findfile_:

```python
file_info = gs.find_file(map_name, element="raster")
full_name = file_info["fullname"]
name = file_info["name"]
mapset = file_info["mapset"]
```

#### Messages

For any informational output, use the _gs.message_ function or _gs.verbose_. For
error messages, use _gs.fatal_ (ends execution) or _gs.error_ (just prints error).
For warnings, use  _gs.warning_. For debugging purposes use _gs.debug_.

```py
# normal message:
gs.message(_("Done."))

# verbose message:
gs.verbose(_("Computation finished successfully."))

# warning:
gs.warning(_("No input values found, using default values."))

# error:
gs.error(_("No map found."))

# fatal error:
# prints error and exits or raises exception (use set_raise_on_error to set the behavior)
gs.fatal(_("No map found, exiting."))

# debug output (use g.gisenv to enable/disable)
# debug level is 1 to 5 (5 is most detailed)

> gs.debug(_("Our calculated value is: {}.".format(value), 3)

Reviewer 2: Missing closing parenthesis.

```py
gs.debug(_("Our calculated value is: {}.").format(value), 3)
```

#### Use Standard Options in Interface

GRASS tools must use the GRASS parser to handle its command line parameters. To
make writing parameters simpler and the interfaces more unified, use standard
options. See
[Parser standard options](https://grass.osgeo.org/grass82/manuals/parser_standard_options.html).
For example, use this:

```python
# %option G_OPT_V_INPUT
# %end
# %option G_OPT_R_OUTPUT
# %end
```

If needed, override values which need to be different:

```python
# %option G_OPT_V_INPUT
# % key: point_input
# % label: Name of input vector map with points
# % description: Points used for sampling the raster input
# %end
# %option G_OPT_R_OUTPUT
# % key: raster_input
# % label: Name of sampled raster map
# % description: Raster map which will be sampled by the points
# %end
```

Do not repeat the values when a standard option defines them.

#### Consider both Flags and Options to modify behavior

Flags are boolean options that default to false. Their names
are only one character. They are defined using:

```python
# %flag
# % key: n
# % description: Consider zeros to be null values
# %end
```

On the command line, the flag is used with dash as `-n`. In Python, the flag

> would be used in the _flags_ parameter of `run_command``:

Reviewer: ```suggestion
would be used in the _flags_ parameter of `run_command`:
```
Was the double backtick intented?

> would be used in the _flags_ parameter of `run_command``:

Reviewer: ```suggestion
would be used in the _flags_ parameter of `run_command`:
```

Was the double backtick intended?

this example, an option named `format` can have default value `plain` and `json`
for JSON output. When you later add YAML, you simply add `yaml` to the possible
values without a need for additional options or flags. The interface definition
for the example would look like:

```python
# %option
# % key: format
# % type: string
# % required: yes
# % options: plain,json,yaml
# % label: Output format
# % descriptions: plain;Plain text output;json;JSON output;yaml;YAML output
# % answer: plain
# %end
```

#### Adding description and keywords

Each tool needs to have a description and at least 3 keywords:

<!-- markdownlint-disable line-length -->

```python
# %module
# % label: Generates a raster map using gaussian random number generator.
# % description: Mean and standard deviation of gaussian deviates can be expressed by the user.
# % keyword: raster
# % keyword: surface
# % keyword: random
# %end
```

<!-- markdownlint-enable line-length -->

Notes:

- the **first** keyword is the module family which goes to the
  [module family index](https://grass.osgeo.org/grass-stable/manuals/general.html)
  in the manual
- the **second** keyword is the overall topic which goes to the

>   [topic index](https://grass.osgeo.org/grass-stable/manuals/topics.html) in the

Reviewer: Stable or devel for contributing guidelines?

Author: Good point

  manual
- the **third** (and more) keyword goes to the
  [keyword index](https://grass.osgeo.org/grass-stable/manuals/keywords.html) in
  the manual

These index manual pages are autogenerated during the build process of GRASS
GIS.

#### Lazy import of optional dependencies

For optional dependencies, import only after the _gs.parser_ call. In that way the
tool can be safely compiled even if the dependency is not installed.

```python
def main():
    options, flags = gs.parser()
    try:
        import pandas as pd  # noqa: E402
    except ModuleNotFoundError:
      gs.fatal(_("Pandas library not installed"))
```

#### Tool name

Try to use names which describe shortly the intended purpose of the tool.

The first letters for the tool name should be:

```text
d.    - display commands
db.   - database commands
g.    - general GIS management commands
i.    - imagery commands
m.    - miscellaneous tool commands
ps.   - postscript commands
r.    - raster commands
r3.   - raster3D commands
v.    - vector commands

> t.    - temporal commands

Reviewer 2: Missing wxPython command category:

```text
wx.    - wxPython gui commands
```

Reviewer 2: > we also have g.gui commands, I am not sure what is the difference

Yes `g.gui` name is right name.

```python
gs.vector_history(output)
```

For rasters:

```python
gs.raster_history(output, overwrite=True)
```

### Developing GRASS GUI

Follow
[wxPython style guide](https://wiki.wxpython.org/wxPython%20Style%20Guide).

Please use the following docstring template:

```py
"""!
@package dir.example

@brief Short example package description

Classes:
 - example::ExampleClass

(C) 2024 by the GRASS Development Team

This program is free software under the GNU General Public License
(>=v2). Read the file COPYING that comes with GRASS for details.

@author First Author <first somewhere.com>
@author Second Author <second somewhere.com>
@author Some Other <third somewhere.com> (some particular change)
"""
```

#### Translations

To enable translating of messages to other languages, use full strings, e.g.
(good example):

```python
if ...:
    win.SetLabel(_("Name for new 3D raster map to create"))
else:
    win.SetLabel(_("Name for new raster map to create"))
```

instead of constructing string from several parts (bad example):

```python
if ...:
    maplabel = 'raster map'
else:
    maplabel = '3D raster map'
win.SetLabel(_("Name for new {} to create").format(maplabel))

> ```

Reviewer: When reading great recent docs in other softwares, I often see that the bad examples have comments inside the code block, or even inline, mentionnons it is bad and to not use, in addition to having it said in the docs. Probably because people don't always read and copy code blocks, or for scraping/AI purposes. Sometimes with a good formatting the lines are in red background, but I don't think simple markdown code blocks allow it.

it finds any problem it will abort the commit and try to solve it automatically.
In that case review the changes and run again `git add` and
`git commit`.

It is also possible to run pre-commit manually, e.g:

```bash
pre-commit run clang-format --all-files
pre-commit run black --all-files
```

Or to target a specific set of files:

```bash
pre-commit run --files raster/r.somemodule/*
```

The pre-commit hooks set is defined in
[.pre-commit-config.yaml](../../.pre-commit-config.yaml).

It is possible to temporally disable the pre-commit hooks in the repo, e.g. while
working on older branches:

```bash
# backporting...
pre-commit uninstall
```

And to reactivate pre-commit again:

```bash
git switch main
pre-commit install
```

### Documentation

There are three types of documentation: C API, Python API and tool documentation.

#### C API documentation

We
[​use doxygen and document the functions](https://grass.osgeo.org/programming8/)
directly in the source code. See lib/gis/\*.c and lib/gis/gislib.dox for
examples.

#### Python API documentation

Python API documentation is written in reStructuredText (reST) which is
compiled with Sphinx (see

> [​PyGRASS documentation](https://grass.osgeo.org/grass-devel/manuals/libpython/pygrass_index.html))

Reviewer 3: ```suggestion
[grass package documentation](https://grass.osgeo.org/grass84/manuals/libpython/))
```

```python
def func(arg1, arg2):
    """Summary line.

    Extended description of function.

    :param int arg1: Description of arg1.
    :param str arg2: Description of arg2.
    :raise: ValueError if arg1 is equal to arg2
    :return: Description of return value
    :rtype: bool

>     :example:

Reviewer 3: We actually never use `:example:`, let's go with more plain `Example:`

```suggestion
    Example:
```

Documentation of a tool should come with clear descriptions, hints on the
implemented algorithm and example(s) with figures.

Each tool (also called "module") comes with its own manual page written in
simple HTML. The HTML file contains **no header nor footer**. The complete HTML
file is autogenerated during the compilation process (with ``--html-description
parameter``). In order to make sure that manuals build without issues, it can be
useful to validate the HTML before submitting, e.g. with
[validator.w3.org](https://validator.w3.org/).

Name the documentation file `'<tool>.html'`, e.g., if the tool is named
r.example, the documentation file should be named `r.example.html`.

##### Markup style guide

The structure consists of several required and optional sections:

```html
<h2>DESCRIPTION</h2>
<!-- required -->

<h2>NOTES</h2>
<!-- suggested -->

<h2>EXAMPLES</h2>
<!-- suggested -->

<h2>TODO</h2>
<!-- optional -->

<h2>KNOWN ISSUES</h2>
<!-- optional -->

<h2>REFERENCES</h2>
<!-- optional -->

<h2>SEE ALSO</h2>
<!-- required -->

<h2>AUTHORS</h2>
<!-- required -->
```

Sections _Notes_, _Examples_, _References_, and _Authors_ can be also in
singular form (e.g, _Note_).

> Note that HTML is converted to MAN pages by

Reviewer 3: ```suggestion
Note that HTML is converted to man pages by
```

> [tools/g.html2man/](https://trac.osgeo.org/grass/browser/grass/trunk/tools/g.html2man).

Reviewer 3: ```suggestion
[utils/g.html2man/](../../utils/g.html2man/).
```

> Since the MAN conversion is limited, please use no other HTML tags than:

Reviewer 3: ```suggestion
Since the man conversion is limited, please use no other HTML tags than:
```

```html
<a> <b> <body> <br> <code> <dd> <dl> <dt> <em>
<h2> <h3> <h4> <head> <hr> <i> <img> <li> <ol> <p>
<pre> <sup> <table> <td> <th> <title> <tr> <ul>
```

More notes on markup:

- Tool names (i.e., v.category) should be emphasized with `<em>v.category</em>`.
- Flags and parameter names written in boldface like `<b>-f</b>` and
  `<b>input</b>`.
- Shell commands, names, values, etc. should use `<tt>42</tt>`.
- Emphasized phrases `<i>should use italics</i>`.
- In order to minimize potential git merge conflicts, please break a line at
  approximately 70-80 chars.

Examples should be coded like this:

```html
<div class="code">

>   <pre>

Reviewer 3: ```suggestion
<pre>
```

## Code Style and Formatting

### Python

We follow the [PEP8](http://www.python.org/dev/peps/pep-0008/) style guide for
Python. Docstrings follow [PEP257](http://www.python.org/dev/peps/pep-0257/) and
use
[Sphinx style](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html).

We use the following tools in the continuous integration to ensure compliance
with PEP8, consistent formatting, and readable and error-free code:

- [flake8](https://flake8.pycqa.org/en/latest/) tool to ensure compliance with
  PEP8
- [black](https://black.readthedocs.io/en/stable/) code formatter to ensure
  consistent code formatting
- [pylint](https://pylint.readthedocs.io/en/latest/) static code analyser to
  find bad code practices or errors

Note that while the entire GRASS code base is black formatted, full complience
with PEP8 is still work in progress.

It is highly recommended to install and use [pre-commit](https://pre-commit.com)
before submitting any new or modified code or any other content. The pre-commit
Git hooks set checks validity and executes automated formatting. See
[using pre-commit](#using-pre-commit) for pre-commit setup and usage.

#### Black

> Use Black to format files:

Author: I've seen both, I changed it to Black, Flake8, Pylint

when appropriate but usually it is not!). Vector graphics should be included in
pages as raster images (i.e. PNGs) for portability but the original format
(preferably SVG) should be committed to the repository as well.

Adding the image to the HTML page (r.viewshed example, the screenshot is shown
with a width of 600 pixel but it is clickable in the manual page). If a larger
image is displayed as shrunk, both **width** and **height** HTML parameters
(values must be calculated according to the picture size!) should be set:

```html
<div align="center" style="margin: 10px">
  <a href="r_viewshed.png">
    <img
      src="r_viewshed.png"
      width="600"
      height="600"
      alt="r.viewshed example"
      border="0"
    /> </a
  ><br />
  <i>Figure: Viewshed shown on shaded terrain (observer position in the
    north-east quadrant with white dot; 5m above ground)</i>
</div>
```

## Best Practices

### General

#### Computational Region

Tools typically **do not change the computational region based on the input
data**. Raster processing tools should respect the current computational region.
Vector processing tools may use the current computational region, e.g., to

> selected subset of the input data.

Author: I removed it

#### Consider both Flags and Options to modify behavior

Flags are like options which are booleans with default false with names which
are only one character. They are defined using:

```python
# %flag
# % key: n
# % description: Consider zeros to be null values
# %end
```

On the command line, the flag is used with dash as `-n`. In Python, the flag
would be used in the _flags_ parameter of `run_command``:

```python
gs.run_command(..., flags="n", ...)
```

However, options are often better because they improve readability, clarify the
default behavior, and allow for extension of the interface.

**Example:** Consider a tool which produces text output which by default
produces human-readable plain text output. Then you add JSON output which is
enabled by a flag `j`. Later, you decide to add YAML output. This now needs to
be flag y which needs to be exclusive with flag `j`. Soon, you have several
related flags each exclusive with all the others. Using an option instead of
flag from the beginning allows the interface to accommodate more formats. In
this example, an option named `format` can have default value `plain` and `json`

> from JSON output. When you later add YAML, you simply add `yaml` to possible

Reviewer 4: ```suggestion
for JSON output. When you later add YAML, you simply add `yaml` to the possible
```

# % label: Output format
# % descriptions: plain;Plain text output;json;JSON output;yaml;YAML output
# % answer: plain
# %end
```

#### Adding description and keywords

Each tool needs to have a description and at least 3 keywords:

<!-- markdownlint-disable line-length -->

```python
# %module
# % label: Generates a raster map using gaussian random number generator.
# % description: Mean and standard deviation of gaussian deviates can be expressed by the user.
# % keyword: raster
# % keyword: surface
# % keyword: random
# %end
```

<!-- markdownlint-enable line-length -->

Notes:

- the **first** keyword is the module family which goes to the
  [module family index](https://grass.osgeo.org/grass-stable/manuals/general.html)
  in the manual
- the **second** keyword is the overall topic which goes to the
  [topic index](https://grass.osgeo.org/grass-stable/manuals/topics.html) in the
  manual
- the **third** (and more) keyword goes to the
  [keyword index](https://grass.osgeo.org/grass-stable/manuals/keywords.html) in
  the manual

These index manual pages are autogenerated during the build process of GRASS
GIS.

#### Lazy import of optional dependencies

> For optional dependencies, import only after the _gs.parser_ call. That way the

Reviewer 4: ```suggestion
For optional dependencies, import only after the _gs.parser_ call. In that way the
```

tool can be safely compiled even if the dependency is not installed.

```python
def main():
    options, flags = gs.parser()
    try:
        import pandas as pd  # noqa: E402
    except ModuleNotFoundError:
      gs.fatal(_("Pandas library not installed"))
```

#### Tool name

Try to use names which describe shortly the intended purpose of the tool.

The first letters for the tool name should be:

```text
d.    - display commands
db.   - database commands
g.    - general GIS management commands
i.    - imagery commands
m.    - miscellaneous tool commands
ps.   - postscript commands
r.    - raster commands
r3.   - raster3D commands

> v.    - vector commands

Reviewer 4: ```suggestion
t.    - temporal commands
v.    - vector commands
```
